text/0000-capture-disjoint-fields.md

```diff
@@ -0,0 +1,196 @@
+        // Error! can't destructure and mutate same data
+        self.0.into_iter()
+            .filter(|s| &s == &self.0[0])
+            .count()
+    }
+
+    pub fn remove_first(&mut self) {
+        // Error! can't mutate and inspect same data
+        self.0.retain(|s| s != &self.0[0])
+    }
+}
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+There exists an nonoptimal but trivial desugar/workaround that covers all cases
+via the following expansion:
+```
+capture := [|'&'|'&mut '] ident ['.' ident]*
+
+'|' args '|' [$e($c:capture):expression]* =>
+'{'
+    ['let' $name:ident '=' $c ';']*
+    '|' args '|' [$e($name)]*
+'}'
+```
+
+Applied to the first two examples:
+```rust
+let _a = &mut foo.a;
+let b = &mut foo.b;
+|| b;
+```
+```rust
+let _a = &mut foo.a;
+let b = foo.b;
+move || b;
+```
+
+This proves that the RFC can be safely implemented without violating any
+existing assumptions. Also, because the compiler would become strictly more
+lenient, it is nonbreaking.
+
+This RFC should not be implemented as such a desugar. Rather, the two following
+changes might be made:
+- Borrowck rules are altered so that capture-by-reference allows simultaneous
+  borrowing of disjoint fields.
+- Codegen and borrowck are altered so that move closures capture only used
+  fields.
+
+## Examples of an ideal implementation
+
+```rust
+|| &mut foo.a;
+```
+- Borrowck passes because `foo` is not borrowed elsewhere.
+- The closure captures a pointer to `foo.a`.
+
+```rust
+let _a = &mut foo.a;
+|| &mut foo.b;
+```
+- Borrowck passes because `foo.a` and `foo.b` are disjoint.
+- The closure captures a pointer to `foo.b`.
+
+```rust
+let _a = &mut foo.a;
+|| (&mut foo.b, &mut foo.c);
+```
+- Borrowck passes because `foo.a`, `foo.b`, and `foo.c` are disjoint.
+- The closure captures a pointer to `foo`.
+
+```rust
+move || foo.a;
+```
+- Borrowck passes because `foo` is not borrowed elsewhere.
+- The closure moves and captures `foo.a` but not `foo.b`.
+
+```rust
+let _a = &mut foo.a;
+move || foo.b;
+```
+- Borrowck passes because `foo.a` and `foo.b` are disjoint.
+- The closure moves and captures `foo.b`.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+This RFC does ruin the intuition that *all* variables named within a closure are
+captured. I argue that that intuition is not really necessary, and killing it is
+worth the cost.
+
+# Rationale and alternatives
+[alternatives]: #alternatives
+
+This proposal is purely ergonomic since there is a complete and common
+workaround. The existing rules could remain in place and rust users could
+continue to pre-borrow/move fields. However, this workaround results in
+significant boilerplate when borrowing many but not all of the fields in a
+struct. It also produces a larger closure than necessary which could be the
+difference between inlining and heap allocation.
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+- How exactly does codegen change?
```

Comment:
If the struct is *not* used outside the closure, this will add one reference per field used in the closure instead of just one reference for the entire struct

Reply:
@oli-obk That would be the trivial implementation, but (as per previous discussion) the closure would be smaller if it captured a single nonexclusive pointer to the whole struct and then derefrenced with a constant offsets.